### PR TITLE
Export table page object from nimble-angular

### PR DESCRIPTION
**Pull request by mollykreis**
# Pull Request

## 🤨 Rationale

The primary driver of this change is being able to access the `TablePageObject` from tests within an Angular client. However, we decided that now is a good time to add secondary entry points for table related code. Creating secondary entry points for everything within nimble-angular is covered by #172.

## 👩‍💻 Implementation

nimble-angular now has the following entry points:
- `@ni/nimble-angular` - this is the existing entry point and contains all the non-table components
- `@ni/nimble-angular/internal-utilities` - contains template-helper code; creating an additional entry point was the best way I could find to share code between other entry points
- `@ni/nimble-angular/table` - contains nimble-table
- `@ni/nimble-angular/table/testing` - exports the table's page object
- `@ni/nimble-angular/table-column` - contains the base class for column directives
- `@ni/nimble-angular/table-column/anchor` - contains nimble-table-column-anchor
- `@ni/nimble-angular/table-column/text` - contains nimble-table-column-text

## 🧪 Testing

- Built systemlink-lib-angular with these changes to verify that it can still import everything it needs, including the nimble-table and nimble-table-column-text
- Used a locally build npm package to write some tests using the page object in systemlink-lib-angular.

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: angular-workspace/projects/ni/nimble-angular/src/directives/table/nimble-table.directive.ts [hunk 1]
```diff
@@ -7,6 +7,8 @@ import type { Observable, Subscription } from 'rxjs';
 export type { Table };
 export type { TableActionMenuToggleEventDetail, TableRowSelectionEventDetail };
 export { TableRecord, TableFieldName, TableFieldValue, TableValidity, TableRowSelectionMode };
+// eslint-disable-next-line no-restricted-imports
+export * from '@ni/nimble-components/dist/esm/table/testing/table.pageobject';
```
Comment (rajsite) [inline angular-workspace/projects/ni/nimble-angular/src/directives/table/nimble-table.directive.ts:11]: I don't think it makes sense to export from the directive. The idea would be to have separate `@ni/nimble-angular/table` and `@ni/nimble-angular/table/testing` entry points.

Hmm actually I don't think there is any reason to not try that with the table now?
What do you think about moving the table to that layout?

https://github.com/ng-packagr/ng-packagr/blob/main/docs/secondary-entrypoints.md

File: angular-workspace/projects/ni/nimble-angular/src/directives/table/nimble-table.directive.ts [hunk 2]
```diff
@@ -7,6 +7,8 @@ import type { Observable, Subscription } from 'rxjs';
 export type { Table };
 export type { TableActionMenuToggleEventDetail, TableRowSelectionEventDetail };
 export { TableRecord, TableFieldName, TableFieldValue, TableValidity, TableRowSelectionMode };
+// eslint-disable-next-line no-restricted-imports
```
Comment (jattasNI) [inline angular-workspace/projects/ni/nimble-angular/src/directives/table/nimble-table.directive.ts:10]: A few thoughts:
1. Could we export this from a separate file with `testing` in its path? That way clients could set up their own `no-restricted-imports` rules to ban imports of testing code from production code. And instead of disabling `no-restricted-imports` here, we could change the [angular-workspace eslint configuration](https://github.com/ni/nimble/blob/874ed79d0d1c0f8228f72de3d12cf562e89d8ada/angular-workspace/.eslintrc.js#L37) to say that imports from testing paths are allowed if the path contains testing. (could probably just update the existing `*.spec.ts` pattern to include `*testing*` or something like that)
2. Once this is exported, can we remove [the suppression in the anchor column directive spec](https://github.com/ni/nimble/blob/874ed79d0d1c0f8228f72de3d12cf562e89d8ada/angular-workspace/projects/ni/nimble-angular/src/directives/table-column/anchor/tests/nimble-table-column-anchor-router-link-with-href.directive.spec.ts#L8)?

File: angular-workspace/tsconfig.json [hunk 3]
```diff
@@ -5,6 +5,24 @@
     "paths": {
       "@ni/nimble-angular": [
         "dist/ni/nimble-angular"
+      ],
```
Comment (rajsite) [inline angular-workspace/tsconfig.json:8]: tsconfig paths supports wildcards, I think we just need:

```json
"paths": {
      "@ni/nimble-angular": [
        "dist/ni/nimble-angular"
      ],
      "@ni/nimble-angular/*": [
        "dist/ni/nimble-angular/*"
      ],
}
```

Comment (mollykreis) [reply angular-workspace/tsconfig.json:8]: Yes, you are correct. That makes this configuration much better!

File: angular-workspace/projects/ni/nimble-angular/src/directives/table/nimble-table.directive.ts (hunk 2 above)
Comment (mollykreis) [reply angular-workspace/projects/ni/nimble-angular/src/directives/table/nimble-table.directive.ts]: I created entry points for the table, each column, and table testing. I believe everything you've outlined in your comment is addressed by that change (including updating the `testing` directories from being able to use `testing` code from nimble-components).

File: angular-workspace/projects/ni/nimble-angular/table-column/anchor/nimble-table-column-anchor.directive.ts [hunk 4]
```diff
@@ -1,9 +1,10 @@
 import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
 import type { TableColumnAnchor } from '@ni/nimble-components/dist/esm/table-column/anchor';
-import type { AnchorAppearance } from '../../anchor/nimble-anchor.directive';
-import { BooleanValueOrAttribute, NumberValueOrAttribute, toBooleanProperty, toNullableNumberProperty } from '../../utilities/template-value-helpers';
+import { AnchorAppearance } from '@ni/nimble-components/dist/esm/anchor/types';
```
Comment (jattasNI) [inline angular-workspace/projects/ni/nimble-angular/table-column/anchor/nimble-table-column-anchor.directive.ts:3]: I'm trying to understand what's going on here. Previously it was `nimble-anchor.directive` that re-exported the `AnchorAppearance` type from `nimble-components`. And `nimble-table-column-anchor.directive` was able to import it from there. But now `nimble-table-column-anchor.directive` is using the type from `nimble-components` and also re-exporting it.

Is this change necessary because this can't refer to the type from the `nimble-anchor.directive` in the main entry point? Or it could, but we like this pattern better because it means that clients can import everything they need from this entry point rather than having to import the type from one entry point and the directive from another?

**Review by jattasNI — APPROVED** [review]

File: angular-workspace/projects/ni/nimble-angular/table-column/anchor/nimble-table-column-anchor.directive.ts (hunk 4 above)
Comment (mollykreis) [reply angular-workspace/projects/ni/nimble-angular/table-column/anchor/nimble-table-column-anchor.directive.ts:3]: I'm pretty sure the previous code was incorrect -- the anchor column should not have been depending on the anchor directive to export a type that is needed to use the column. When the type was needed by two directives (the anchor directive and the anchor column directive), the export of the type should have been moved to `public-api.ts` for the main entry point. We have [a section in that file](https://github.com/ni/nimble/blob/980eb3cc9d18d2c8c855f784180a6fe1925947a3/angular-workspace/projects/ni/nimble-angular/src/public-api.ts#L112) that exports types that are needed by multiple components.

To summarize & answer some of your specific questions -- even if we could refer to the type from `nimble-anchor.directive` from here (which we can't), that wouldn't be the correct approach. The pattern introduced here is preferred because clients an import everything they need to use a component from a single entry point.

Comment (jattasNI) [reply angular-workspace/projects/ni/nimble-angular/table-column/anchor/nimble-table-column-anchor.directive.ts:3]: Makes sense, thanks!

**Review by atmgrifter00 — APPROVED** [review]

File: change/@ni-nimble-angular-cef992f7-0e96-4b96-8c49-907a8a794bd0.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (rajsite) [inline change/@ni-nimble-angular-cef992f7-0e96-4b96-8c49-907a8a794bd0.json:1]: I like a lot of how this is going. My only concern is that taken to the end this pattern results in pretty much nothing in the src folder and a lot of top-level folders. So I think I'm on the boat of keeping all the source inside the top-level `src` folder.

So it'll look like:
```
/src/directives - legacy stuff to make into new entrypoints
/src/table-column - new entrypoint
/src/table - new entrypoint
/src/internal-utilities - new entrypoint
```
But I don't think we need to change anything about the structure inside the `table-column/`, `table/`, and `internal-utilities` folders as you have shown them.

Maybes it's worthwhile to sync on this idea with @atmgrifter00 and @jattasNI to avoid extra churn on the direction?

File: angular-workspace/projects/ni/nimble-angular/table/testing/ng-package.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
+    "$schema": "../../../../../../node_modules/ng-packagr/ng-package.schema.json",
```
Comment (rajsite) [inline angular-workspace/projects/ni/nimble-angular/table/testing/ng-package.json:2]: Just to double check, do we need to specify these schema attributes? it's a bit of a bummer that we need to link out to the workspace root but not a deal breaker.

Comment (rajsite) [reply angular-workspace/projects/ni/nimble-angular/table/testing/ng-package.json:2]: sounds handy enough then 👍

File: change/@ni-nimble-angular-cef992f7-0e96-4b96-8c49-907a8a794bd0.json (hunk 5 above)
Comment (mollykreis) [reply change/@ni-nimble-angular-cef992f7-0e96-4b96-8c49-907a8a794bd0.json:1]: From what I can tell the directory structure you are suggesting isn't supported. The entry point name is automatically determined from the directory hierarchy containing the `ng-package.json` file. Therefore, with the structure you're suggesting, our entry points would be `@ni/nimble-angular/src/table-column`, `@ni/nimble-angular/src/table`, etc. I don't think that's what we want.

I found an issue with a lot of discussion around this topic here: https://github.com/ng-packagr/ng-packagr/issues/900

Comment (rajsite) [reply change/@ni-nimble-angular-cef992f7-0e96-4b96-8c49-907a8a794bd0.json:1]: Ah didn't realize that wasn't straightforward. I don't have any deal breaking concerns with the current approach and the public api of it looks good. We can definitely start this way! Thanks for looking into it 👍

**Review by rajsite — APPROVED** [review]
